pkg/api/types.go

```diff
@@ -1414,6 +1414,38 @@ type ReplicationControllerList struct {
 	Items []ReplicationController `json:"items"`
 }
 
+// DisruptionBudget is an object to define the max disruption that can be caused to a collection of pods
+type DisruptionBudget struct {
+	unversioned.TypeMeta `json:",inline"`
+	ObjectMeta           `json:"metadata,omitempty"`
+
+	// Specification of the desired behavior of the DisruptionBudget.
+	Spec DisruptionBudgetSpec `json:"spec,omitempty"`
+	// Most recently observed status of the DisruptionBudget.
+	Status DisruptionBudgetStatus `json:"status,omitempty"`
+}
+
+// DisruptionBudgetSpec is a description of a DisruptionBudget.
+type DisruptionBudgetSpec struct {
+	// The minimum number of pods that must be available simultaneously.  Exactly
```

Comment:
Regarding your last comment: That's a really great point. As you alluded to, to compute the denominator for the percentage, you'd need to iterate through all controllers in the system, and add up the number of replicas in the "intended state" of all controllers' specs whose pods use this DisruptionBudget. But in fact that would still be wrong, because in theory you can have pods with no controller, that point back to a DisruptionBudget. The denominator is fundamentally not well-defined.

I don't think it works to use the number of pods you see as the "expected" number of pods, because then by definition 100% will always be available. (Though maybe I misunderstood that part of your comment.)

I guess we should just do MinAvailable (an int), and expect that the config system would have templating that would allow you to say "MinAvailable = 0.8 * X" (e.g. you use X in your PodTemplate to set number of replicas) and the value would get filled in when the config is evaluated...